File: csharp/pack/packable/LongArrayPool.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace pack.packable
{
    class LongArrayPool
    {
        private const int DEFAULT_SIZE = 32;
        private const int DEFAULT_CAPACITY = 16;
        private static int defaultCount = 0;
        private static readonly ulong[][] defaultArrays = new ulong[DEFAULT_CAPACITY][];

        private const int SECOND_SIZE = 64;
        private const int SECOND_CAPACITY = 8;
        private static int seondCount = 0;
        private static readonly ulong[][] secondArrays = new ulong[SECOND_CAPACITY][];

        internal static ulong[] GetArray(int size)
        {
            if (size > TagFormat.MAX_INDEX_BOUND)
            {
                // should never happen
                throw new ArgumentOutOfRangeException("request size must less than:" + TagFormat.MAX_INDEX_BOUND);
            }

            if (size <= DEFAULT_SIZE)
            {
                return GetDefaultArray();
            }
            else if (size <= SECOND_SIZE)
            {
                return GetSecondArray();
            }
            else
            {
                return new ulong[(size <= 128) ? 128 : 256];
            }
        }

        internal static void RecycleArray(ulong[] a)
        {
            if (a == null)
            {
                return;
            }
            int size = a.Length;
            if (size == DEFAULT_SIZE)
            {
                RecycleDefaultArray(a);
            }
            else if (size == SECOND_SIZE)
            {
                RecycleSecondArray(a);
            }
            // else, drop it
        }

       internal static ulong[] GetDefaultArray()
        {
            lock (defaultArrays)
            {
                if (defaultCount > 0)
                {
                    ulong[] a = defaultArrays[--defaultCount];
                    defaultArrays[defaultCount] = null;
                    return a;
                }
            }
            return new ulong[DEFAULT_SIZE];
        }

        private static void RecycleDefaultArray(ulong[] a)
        {
            lock (defaultArrays)
            {
                if (defaultCount < DEFAULT_CAPACITY)
                {
                    defaultArrays[defaultCount++] = a;
                }
            }
        }

        private static ulong[] GetSecondArray()
        {
            lock (secondArrays)
            {
                if (seondCount > 0)
                {
                    ulong[] a = secondArrays[--seondCount];
                    secondArrays[seondCount] = null;
                    return a;
                }
            }
            return new ulong[SECOND_SIZE];
        }

        private static void RecycleSecondArray(ulong[] a)
        {
            lock (secondArrays)
            {
                if (seondCount < SECOND_CAPACITY)
                {
                    secondArrays[seondCount++] = a;
                }
            }
        }
    }
}

File: java/packable/src/main/java/io/packable/LongArrayPool.java
package io.packable;

/**
 * LongArrayPool for {@link PackDecoder}
 */
class LongArrayPool {
    private static final int DEFAULT_SIZE = 32;
    private static final int DEFAULT_CAPACITY = 16;
    private static int defaultCount = 0;
    private static final long[][] defaultArrays = new long[DEFAULT_CAPACITY][];

    private static final int SECOND_SIZE = 64;
    private static final int SECOND_CAPACITY = 4;
    private static int secondCount = 0;
    private static final long[][] secondArrays = new long[SECOND_CAPACITY][];

    static long[] getArray(int size) {
        if (size <= DEFAULT_SIZE) {
            return getDefaultArray();
        } else if (size <= SECOND_SIZE) {
            return getSecondArray();
        } else {
            return new long[(size <= 128) ? 128 : 256];
        }
    }

    static void recycleArray(long[] a) {
        if (a == null) {
            return;
        }
        int size = a.length;
        if (size == DEFAULT_SIZE) {
            recycleDefaultArray(a);
        } else if (size == SECOND_SIZE) {
            recycleSecondArray(a);
        }
        // else, drop it
    }

    static long[] getDefaultArray() {
        synchronized (defaultArrays) {
            if (defaultCount > 0) {
                long[] a = defaultArrays[--defaultCount];
                defaultArrays[defaultCount] = null;
                return a;
            }
        }
        return new long[DEFAULT_SIZE];
    }

    private static void recycleDefaultArray(long[] a) {
        synchronized (defaultArrays) {
            if (defaultCount < DEFAULT_CAPACITY) {
                defaultArrays[defaultCount++] = a;
            }
        }
    }

    private static long[] getSecondArray() {
        synchronized (secondArrays) {
            if (secondCount > 0) {
                long[] a = secondArrays[--secondCount];
                secondArrays[secondCount] = null;
                return a;
            }
        }
        return new long[SECOND_SIZE];
    }

    private static void recycleSecondArray(long[] a) {
        synchronized (secondArrays) {
            if (secondCount < SECOND_CAPACITY) {
                secondArrays[secondCount++] = a;
            }
        }
    }
}
